source/auth/auth.rst

```diff
@@ -683,6 +683,212 @@ mechanism
+sts.amazonaws.com us-east-1
+first.second      second
+first             us-east-1
+================= ========= 
+
+`MongoCredential`_ Properties
+`````````````````````````````
+
+username
+	MAY be specified. The non-sensitive IAM access key.
+
+source
+	MUST be "$external". Defaults to ``$external``.
+
+password
+	MAY be specified. The sensitive IAM secret key.
+
+mechanism
+	MUST be "MONGODB-IAM"
+
+mechanism_properties
+	AWS_SESSION_TOKEN
+		Drivers MUST allow the user to specify an AWS session token for authentication with temporary credentials.
+
+IAM Credentials
+   If a username and password is provided drivers MUST use these for the IAM access key and IAM secret key, respectively. Drivers
+   MUST use this secret key to generate a signature using HMAC-SHA256.
+
+Temporary Credentials
+   If a username and password is not provided drivers MUST query the standard local AWS endpoint for temporary credentials.
+   If temporary credentials cannot be obtained then drivers MUST fail authentication and raise an error.
+
+   EC2 Instance
+      Calling the endpoint ``http://169.254.169.254/latest/meta-data/iam/security-credentials/`` from an EC2 instance 
+      will return the role attached to the EC2 instance in plaintext, if it exist:
+
+      ``role-name``
+
+      ``http://169.254.169.254/latest/meta-data/iam/security-credentials/<role-name>`` will return the following json response: 
+
+      .. code:: javascript
+
+            {
+                "Code": "Success",
+                "LastUpdated" : <date>,
+                "Type": "AWS-HMAC",
+                "AccessKeyId" : <access_key>,
+                "SecretAccessKey": <secret_access_key>,
+                "Token" : <security_token>,
+                "Expiration": <date>
+            }
+
+      .. note::
+
+      The IP address 169.254.169.254 is a link-local address and is valid only from the instance. 
+      See `Instance Metadata and User Data <https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html>`_.
+
+      From this response, drivers will need to extract the ``AccessKeyId``, ``SecretAccessKey`` and the ``Token`` argument values. 
+      Drivers will use these values to construct the call to ``sts::getCallerIdentity`` as part of the Amazon Signature V4 algorithm. 
+      Drivers who cache the security temporary credentials must check the expiration date before using them. 
+      If the credentials expire, they must refetch them. 
+
+   ECS Tasks
+      ECS temporary credentials can be retrieved by calling a URL at ``http://169.254.170.2/$AWS_CONTAINER_CREDENTIALS_RELATIVE_URI`` 
+      where ``AWS_CONTAINER_CREDENTIALS_RELATIVE_URI`` is an environment variable set by the AWS EC2 container agent.
+
+   .. code:: javascript
+
+      {
+       "AccessKeyId": "ACCESS_KEY_ID",
+       "Expiration": "EXPIRATION_DATE",
+       "RoleArn": "TASK_ROLE_ARN",
+       "SecretAccessKey": "SECRET_ACCESS_KEY",
+       "Token": "SECURITY_TOKEN_STRING"
+      }
+
+
```

Comment:
Can we add a section that lists a few MONGODB-IAM authentication connection strings?

Reply:
The format is the same as other auth mechanisms. Drivers would have to handle an optional user and password, authMechanism=MONGODB-IAM and one new property, AWS_SESSION_TOKEN. I think adding examples might be a bit verbose since the other mechanisms just rely on the 'Connection String Options' section.

Reply:
@bazile-clyde Are you going to update the URI spec or is there some other place the new property gets documented?

Reply:
AFAIK the possible values for the authMechanism key are only documented in the auth spec. The URI spec points to the auth spec. Each auth mechanism documents its possible authMechanismProperties key-values under its corresponding section in the auth spec under the subsection mechanism_properties. This property, "SECURITY_TOKEN_STRING", can be overridden using the auth mechanism property AWS_SESSION_TOKEN in the URI which is listed under mechanism_properties. Perhaps it's confusing that this value is called SECURITY_TOKEN_STRING and the key is called AWS_SESSION_TOKEN?

Reply:
I really think it would be useful to include a few example URIs. Something like this:
```
Example MongoDB URI
^^^^^^^^^^^^^^^^^^^

This section lists a few example MongoDB URIs that configure MONGODB-IAM authentication.

Auth with a AWS_SESSION_TOKEN:
``mongodb://IAM-access-key:IAM-secret-key@host/?authMechanism=MONGODB-IAM&authSource=$external``

Auth with a Temporary AWS_SESSION_TOKEN:
``mongodb://temporary-IAM-access-key@host/?authMechanism=MONGODB-IAM&authSource=$external&authMechanismProperties=AWS_SESSION_TOKEN:<TOKEN_VALUE>``
```

It took me 10 minutes just to figure that out (and I'm still not even sure my examples are correct).

Reply:
Done. Added example URIs for each section.

Reply:
Thanks!